### PR TITLE
examples how to write unit tests for tensorflow

### PR DESCRIPTION
**Pull request by lkuligin**
b/156336122

### CODE REVIEW

**Comment by jaketf** [comment]
@TheMichaelHu could you review this?

**Comment by TheMichaelHu** [comment]
@jaketf yeah, I'd be happy to

File: examples/tensorflow-profiling-examples/example_test.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    def test_shape_multidim(self):
+        batch_size, dim = (4, 15)
+        x = np.ones((batch_size, 7, dim))
+        layer = example.LinearBlock(dim)
+        output = layer(x)
+        self.assertAllEqual(output.shape, (batch_size, 7, dim))
+
+    def test_output(self):
+        dim = 4
+        batch_size = 3
```
Comment (TheMichaelHu) [inline examples/tensorflow-profiling-examples/example_test.py:51]: nit: inconsistent with the other tests where you initialize batch_size and dim on the same line

File: examples/tensorflow-profiling-examples/example_test.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    def test_shape(self):
+        batch_size, dim = (4, 15)
+        x = np.ones((batch_size, dim))
+        layer = example.LinearBlock(dim)
+        output = layer(x)
+        self.assertAllEqual(output.shape, (batch_size, dim))
+
+    def test_shape_multidim(self):
+        batch_size, dim = (4, 15)
```
Comment (TheMichaelHu) [inline examples/tensorflow-profiling-examples/example_test.py:43]: nit: Just a suggestion, I think there's an opportunity here to use `setUp()` to make these tests less repetitive. Something like:

```python
def setUp(self):
    self.batch_size = 1
    self.input_dim = 2
    self.output_dim = 3
    self.layer = example.LinearBlock()
```

File: examples/tensorflow-profiling-examples/example_test.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+class LinearBlockTest(tf.test.TestCase):
+    def test_shape_default(self):
+        x = np.ones((4, 32))
+        layer = example.LinearBlock()
+        output = layer(x)
+        self.assertAllEqual(output.shape, (4, 32))
+
+    def test_shape(self):
+        batch_size, dim = (4, 15)
+        x = np.ones((batch_size, dim))
```
Comment (TheMichaelHu) [inline examples/tensorflow-profiling-examples/example_test.py:37]: nit: I think having `dim` be both the input dimension and output dimension makes this test harder to read. I suggest separate values so the reader knows the input dim has nothing to do with this test. Same thoughts apply to other tests.

File: examples/tensorflow-profiling-examples/example_test.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    @patch.object(initializers, 'get', lambda _: tf.compat.v1.keras.initializers.Ones)
+    def test_output_ones(self):
+        dim = 4
+        batch_size = 3
+        output_dim = 2
+        x = np.ones((batch_size, dim))
+        layer = example.LinearBlock(output_dim)
+        output = layer(x)
+        expected_output = np.ones((batch_size, output_dim)) * (dim + 1)
```
Comment (TheMichaelHu) [inline examples/tensorflow-profiling-examples/example_test.py:66]: nit: this might need a comment or docstring explaining how it's essentially `np.ones((batch_size, dim)).dot(np.ones((dim, output_dim))) + np.ones((batch_size, 1))` because the initializes for `w` and `b` were replaced with ones.

File: examples/tensorflow-profiling-examples/example_test.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        dataset['Origin'] = dataset['Origin'].map({1: 'USA', 2: 'Europe', 3: 'Japan'})
+        dataset = pd.get_dummies(dataset, prefix='', prefix_sep='')
+        dataset = dataset[dataset.columns].astype('float64')
+        labels = dataset.pop('MPG')
+        batch_size =   10
+        df  = pd.DataFrame()
+        LABEL = ""
+        faked_dataset = tf.data.Dataset.from_tensor_slices((dict(df.pop(LABEL)), df[LABEL].values))
+        return faked_dataset.repeat().batch(batch_size)
+
```
Comment (TheMichaelHu) [inline examples/tensorflow-profiling-examples/example_test.py:89]: Are lines 84-89 meant to be here? They look like they were added for debugging or something.

File: examples/tensorflow-profiling-examples/example_test.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        return dataset, labels
+
+    def test_basic(self):
+        train_dataset, train_labels = self._get_data()
+        dim = len(train_dataset.keys())
```
Comment (TheMichaelHu) [inline examples/tensorflow-profiling-examples/example_test.py:94]: nit: `dim = len(train_dataset)` also works. This might be misleading to read, however, unless `train_dataset` is renamed to `train_features`.

File: examples/tensorflow-profiling-examples/example_test.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        example_model = example.get_model(dim)
+        test_ind = train_dataset.sample(10).index
+        test_dataset, test_labels = train_dataset.iloc[test_ind], train_labels.iloc[test_ind]
```
Comment (TheMichaelHu) [inline examples/tensorflow-profiling-examples/example_test.py:97]: haha janky train/test splitting that leaves the test set in the training set. No action needed, just thought it funny.

File: examples/tensorflow-profiling-examples/example_test.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
+import unittest
+from unittest.mock import MagicMock
```
[the remark this answers is not in the document]
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:2]: fixed

File: examples/tensorflow-profiling-examples/example_test.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+import tensorflow as tf
+from tensorflow.python.keras import testing_utils
+from tensorflow.python.keras import initializers
+
+import example
+
+
+class LinearBlockFullTest(tf.test.TestCase):
+    def test_basic(self):
+       testing_utils.layer_test(example.LinearBlockFull, input_shape=(4, 32))
```
[the remark this answers is not in the document]
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:16]: yep, I'm sorry about that, it looks like some last-minute spacing issue

File: examples/tensorflow-profiling-examples/example_test.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        history = example_model.fit(
+            train_dataset, train_labels, steps_per_epoch=2, epochs=2,
+            batch_size=10,
+            validation_split=0.1,
+            validation_steps=1)
+        self.assertAlmostEqual(
+            history.history['mse'][-1],
+            history.history['loss'][-1],
+            places=2)
+        _ = example_model.evaluate(test_dataset, test_labels)
```
[the remark this answers is not in the document]
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:107]: fixed

File: examples/tensorflow-profiling-examples/example.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
+from absl import logging
```
[the remark this answers is not in the document]
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example.py:1]: fixed

File: examples/tensorflow-profiling-examples/example_test.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+                              sep=' ', skipinitialspace=True)
+        dataset = dataset.dropna()
+        dataset['Origin'] = dataset['Origin'].map({1: 'USA', 2: 'Europe', 3: 'Japan'})
+        dataset = pd.get_dummies(dataset, prefix='', prefix_sep='')
+        dataset = dataset[dataset.columns].astype('float64')
+        labels = dataset.pop('MPG')
+        batch_size =   10
+        df  = pd.DataFrame()
+        LABEL = ""
+        faked_dataset = tf.data.Dataset.from_tensor_slices((dict(df.pop(LABEL)), df[LABEL].values))
```
[the remark this answers is not in the document]
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:87]: fixed

File: examples/tensorflow-profiling-examples/example_test.py (hunk 1 above)
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:51]: fixed

File: examples/tensorflow-profiling-examples/example_test.py (hunk 5 above)
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:89]: you're totally right, thanks! fixed.

File: examples/tensorflow-profiling-examples/example_test.py (hunk 7 above)
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:97]: yep, since it's for testing purposes only, it's the easiest way to create a faked dataset actually

File: examples/tensorflow-profiling-examples/example_test.py (hunk 6 above)
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:94]: yep, renamed.
I just wanted to highlight that `train_features` is a dict actually (otherwise it might be misleading too, since you might be expect `len(train_features)` is amount of  training examples.

File: examples/tensorflow-profiling-examples/example_test.py (hunk 3 above)
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:37]: fixed

File: examples/tensorflow-profiling-examples/example_test.py (hunk 2 above)
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/example_test.py:43]: I thought about this initially, but since dims and batch_size are different in each test, and LinearBLock needs to know about output_dim, I couldn't come up with a good idea how setUp can help.

File: README.md [hunk 13]
```diff
@@ -50,6 +50,7 @@ The examples folder contains example solutions across a variety of Google Cloud
 * [QAOA](examples/qaoa) - Examples of parsing a max-SAT problem in a proprietary format.
 * [Redis Cluster on GKE Example](examples/redis-cluster-gke) - Deploying Redis cluster on GKE.
 * [Spinnaker](examples/spinnaker) - Example pipelines for a Canary / Production deployment process.
+* [TensorFlow unit testing](examples/tensorflow-profiling-examples) - How to write unit tests for TensorFlow models.
```
Comment (TheMichaelHu) [inline README.md:53]: I think this should link to `tensorflow-unit-testing`
Also, it looks like the convention is to capitalize all words in the name (i.e. TensorFlow Unit Testing)

File: examples/tensorflow-unit-testing/example_test.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        expected_output = np.ones((batch_size, output_dim)) * (batch_size, input_dim, output_dim + 1)
+        self.assertAllClose(output, expected_output, atol=1e-4)
+
+
+class ExampleModelTest(tf.test.TestCase):
+    def _get_data(self):
+        dataset_path = tf.keras.utils.get_file(
+            'auto-mpg.data',
+            'http://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data')
+        column_names = ['MPG','Cylinders','Displacement','Horsepower','Weight',
```
Comment (TheMichaelHu) [inline examples/tensorflow-unit-testing/example_test.py:84]: nit: Commas should be followed by a space. Surprised your linter didn't catch this.
Some other things my linter has complained about:
- Some lines exceed 80 chars
- Would be nice to have more docstrings.

File: examples/tensorflow-profiling-examples/README.md [hunk 15]
```diff
@@ -1,143 +0,0 @@
-# TensorFlow Profiling Examples
```
[the remark this answers is not in the document]
Comment (lkuligin) [reply examples/tensorflow-profiling-examples/README.md]: yep, it has been fixed already

File: README.md (hunk 13 above)
Comment (lkuligin) [reply README.md:53]: fixed

File: examples/tensorflow-unit-testing/example_test.py (hunk 14 above)
Comment (lkuligin) [reply examples/tensorflow-unit-testing/example_test.py:84]: fixed

File: README.md [hunk 16]
```diff
@@ -77,7 +78,6 @@ The tools folder contains ready-made utilities which can simpilfy Google Cloud P
 * [GCP Organization Hierarchy Viewer](tools/gcp-org-hierarchy-viewer) - A CLI utility for visualizing your organization hierarchy in the terminal.
 * [GCS Bucket Mover](tools/gcs-bucket-mover) - A tool to move user's bucket, including objects, metadata, and ACL, from one project to another.
 * [GCS Usage Recommender](tools/gcs-usage-recommender) - A tool that generates bucket-level intelligence and access patterns across all projects for a GCP project to generate recommended object lifecycle management.
-* [GCS to BigQuery](tools/gcs2bq) - A tool fetches object metadata from all Google Cloud Storage buckets and exports it in a format that can be imported into BigQuery for further analysis.
```
Comment (TheMichaelHu) [inline README.md:80]: Accidental deletion?

Comment (lkuligin) [reply README.md]: Yep, thanks for spotting this!

**Comment by TheMichaelHu** [comment]
@jaketf Can you review and merge, I don't have permission.

File: examples/tensorflow-unit-testing/README.md [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+errors, but in the basic case we don’t check whether the training itself makes sense - i.e., whether
+a loss decreases to any meaningful value. But more about it later.
+Let’s have a look at a simple example of how to test a model from this [tutorial](https://www.tensorflow.org/tutorials/keras/regression).
+```python
+class ExampleModelTest(tf.test.TestCase):
+   def _get_data(self):
+       dataset_path = tf.keras.utils.get_file(
+           'auto-mpg.data',
+           'http://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data')
+       column_names = ['MPG','Cylinders','Displacement','Horsepower','Weight',
```
Comment (TheMichaelHu) [inline examples/tensorflow-unit-testing/README.md:73]: fyi this doesn't match the code after the style fixes were applied

Comment (lkuligin) [reply examples/tensorflow-unit-testing/README.md:73]: Thanks, let's leave as it is for now!
